samcli/lib/build/bundler.py

```diff
@@ -134,6 +143,53 @@ def set_sourcemap_env_from_metadata(self) -> Dict:
 
         return template
 
+    def _check_invalid_lambda_handler(self, handler: str, name: str) -> bool:
+        """
+        Function to check if the handler exists in the build dir where we expect it to.
+        If it does, we won't change the path to prevent introducing breaking changes.
+
+        :param handler: handler string as defined in the template.
+        :param name: function name corresponding to function build directory
+        :return: True if it's an invalid handler, False otherwise
+        """
+        if not self._build_dir:
+            return False
+        handler_filename = self._get_path_and_filename_from_handler(handler)
+        expected_artifact_path = Path(self._build_dir, name, handler_filename)
+        return not expected_artifact_path.is_file()
+
+    @staticmethod
+    def _get_path_and_filename_from_handler(handler: str) -> str:
+        """
+        Takes a string representation of the handler defined in the
+        template, returns the file name and location of the handler.
+
+        :param handler: string representation of handler property
+        :return: string path to built handler file
+        """
+        return handler.split(".")[HANDLER_WITHOUT_FUNCTION_INDEX] + ".js"
+
+    def _update_function_handler(self, template: Dict) -> Dict:
+        """
+        Updates the function handler to point to the actual handler,
+        not the pre-built handler location.
+
+        E.g. pre-build could be codeuri/src/handler/app.lambdaHandler
+        esbuild would bundle that into .aws-sam/FunctionName/app.js
+
+        :param template: deepcopy of template dict
+        :return: Updated template with resolved handler property
+        """
+        for name, resource in self._stack.resources.items():
+            if self._esbuild_in_metadata(resource.get("Metadata", {})):
+                long_path_handler = resource.get("Properties", {}).get("Handler", "")
+                if not long_path_handler or not self._check_invalid_lambda_handler(long_path_handler, name):
+                    continue
+                resolved_handler = long_path_handler.split("/")[HANDLER_FILE_INDEX]
```

Comment:
does it matter on windows?

Reply:
Good question.. this depends on the CFN spec and what's expected in the handler property. I'm not sure if CFN expects Windows file separators but can look into it more tomorrow.

Reply:
File separators for in the template should be the same regardless of OS. For example, our templates don't make a distinction https://github.com/aws/aws-sam-cli-app-templates/blob/master/nodejs18.x/web/%7B%7Bcookiecutter.project_name%7D%7D/template.yaml.